subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/VersionCatalogExtensionIntegrationTest.groovy

```diff
@@ -2306,4 +2306,130 @@ Second: 1.1"""
         expect:
         succeeds ':help'
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/23096")
+    def 'all properties of version catalog dependencies are copied when the dependency is copied'() {
+        given:
+        buildFile << """
+            configurations {
+                implementation
+                destination1
+                destination2
+            }
+
+            dependencies {
+                implementation(libs.test) {
+                    because("reason1")
+
+                    exclude(group: "test-group", module: "test-module")
+                    artifact {
+                        name = "test-name"
+                        classifier = "test-classifier"
+                        extension = "test-ext"
+                        type = "test-type"
+                        url = "test-url"
+                    }
+                    transitive = true
+                    endorseStrictVersions()
+
+                    version {
+                        branch = "branch"
+                        strictly("123")
+                        prefer("789")
+                        reject("aaa")
+                    }
+
+                    changing = true
+                }
+                implementation(libs.test) {
+                    transitive = false
+                    targetConfiguration = "abc"
+                    doNotEndorseStrictVersions()
+
+                    version {
+                        require("456")
+                    }
+
+                    changing = false
+                }
+                implementation(libs.test) {
+                    attributes {
+                        attribute(Attribute.of('foo', String), 'bar')
+                    }
+                    capabilities {
+                        requireCapability("org:test-cap:1.1")
+                    }
+                }
+            }
+
+            def printDep(dep) {
+                // Dependency
+                println("Dependency: " + dep.group + " " + dep.name + " " + dep.version + " " + dep.reason)
+
+                // ModuleDependency
+                dep.excludeRules.each { println("Exclude rule: " + it.group + " " + it.module) }
+                dep.artifacts.each { println("Artifact: " + it.name + " " + it.type + " " + it.extension + " " + it.classifier + " " + it.url) }
+                println("ModuleDependency: " + dep.transitive + " " + dep.targetConfiguration + " " + dep.attributes + " " + dep.requestedCapabilities)
+
+                // ExternalDependency + ExternalModuleDependency
+                println("External: " + dep.isChanging())
+            }
```

Comment:
Instead of relying on text output, could we make these asserts?

IOW, for each configuration we're evaluating the dependencies for, we'd loop through their dependencies and assert a few things:
- We have exactly 3 dependencies
- The dependency has a particular GAV
- The dependency has a particular artifact/version constraint/other properties depending on what should have been set above. 

In the build script, we'd use `assert ...` to confirm these things.

Reply:
Updated to use asserts